akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1252,9 +1195,39 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
     // make sure any invokeWithFeedback after this fails fast
     // and fail current outstanding invokeWithFeedback promises
     val inProgress = asyncCallbacksInProgress.getAndSet(null)
-    inProgress.foreach { case (acb, promises) ⇒ acb.onStop(promises) }
+    if (inProgress.nonEmpty) {
+      val exception = streamDetatchedException
+      inProgress.foreach(_.tryFailure(exception))
+    }
+  }
+
+  private[this] var asyncCleanupCounter = 0L
+
+  /** Called from interpreter thread by GraphInterpreter.runAsyncInput */
+  private[stream] def onFeedbackDispatched(): Unit = {
+    asyncCleanupCounter += 1
+
+    // 256 seemed to be a sweet spot in SendQueueBenchmark.queue benchmarks
+    // It means that at most 255 completed promises are retained per logic that
+    // uses invokeWithFeedback callbacks.
+    //
+    // TODO: add periodical cleanup to get rid of those 255 promises as well
+    if (asyncCleanupCounter % 256 == 0) {
+      @tailrec def cleanup(): Unit = {
+        val previous = asyncCallbacksInProgress.get()
+        if (previous != null) {
+          val updated = previous.filterNot(_.isCompleted)
+          if (!asyncCallbacksInProgress.compareAndSet(previous, updated)) cleanup()
+        }
+      }
+
+      cleanup()
+    }
```

Comment:
What about the case with large amounts of short lived async callbacks, then they will never reach the cleanup limit and will stay in the list until the stage stops?

Reply:
Ah, this isn't in the callback but in the logic, ok!

Reply:
As discussed above already, we might still keep up to 255 completed promises in the list. Did you add a test where we could observe the OOM before? Might make sense to gather a heap dump while it is running and see if it is correct that a Promise will not keep extra references.

Reply:
No I never did add a test, there was a standalone repeater that I deleted though. I'll run that to verify (but I'm pretty convinced that it works as it should).

Reply:
Did some runs with profiler attached as well as taking a heap dump and looking at that, it looks good! :)

Reply:
Cool, thanks.